intake_stac/drivers.py

```diff
@@ -0,0 +1,217 @@
+from intake.source.base import DataSource, PatternMixin, Schema
+from intake.source.utils import reverse_formats
+from pkg_resources import get_distribution
+
+__version__ = get_distribution('intake_stac').version
+
+
+class RioxarraySource(DataSource, PatternMixin):
+    """Open a xarray dataset via Rioxarray.
+    This creates an xarray.DataArray https://github.com/corteva/rioxarray
+
+    Parameters
+    ----------
+    urlpath: str or iterable, location of data
+        May be a local path, or remote path if including a protocol specifier
+        such as ``'s3://'``. May include glob wildcards or format pattern strings.
+        Must be a format supported by rasterIO (normally GeoTiff).
+        Some examples:
+            - ``{{ CATALOG_DIR }}data/RGB.tif``
+            - ``s3://data/*.tif``
+            - ``s3://data/landsat8_band{band}.tif``
+            - ``s3://data/{location}/landsat8_band{band}.tif``
+            - ``{{ CATALOG_DIR }}data/landsat8_{start_date:%Y%m%d}_band{band}.tif``
+    chunks: None or int or dict, optional
+        Chunks is used to load the new dataset into dask
+        arrays. ``chunks={}`` loads the dataset with dask using a single
+        chunk for all arrays. default `None` loads numpy arrays.
+    path_as_pattern: bool or str, optional
+        Whether to treat the path as a pattern (ie. ``data_{field}.tif``)
+        and create new coodinates in the output corresponding to pattern
+        fields. If str, is treated as pattern to match on. Default is True.
+    """
+
+    name = 'rioxarray'
+    version = __version__
+    container = 'xarray'
+    partition_access = True
+
+    def __init__(
+        self,
+        urlpath,
+        chunks=None,
+        concat_dim='concat_dim',
+        override_coords=None,
+        xarray_kwargs=None,
+        metadata=None,
+        path_as_pattern=True,
+        storage_options=None,
+        **kwargs,
+    ):
+        self.path_as_pattern = path_as_pattern
+        self.urlpath = urlpath
+        self.chunks = chunks
+        self.dim = concat_dim
+        # self.storage_options = storage_options or {} #only relevant to fsspec?
+        self.override_coords = override_coords
+        self._kwargs = xarray_kwargs or {}
+        self._ds = None
+        # if isinstance(self.urlpath, list):
+        #    self._can_be_local = fsspec.utils.can_be_local(self.urlpath[0])
+        # else:
+        #    self._can_be_local = fsspec.utils.can_be_local(self.urlpath)
```

Comment:
@martindurant following up on your [comment on the last PR](https://github.com/intake/intake-stac/pull/72#discussion_r585844868) about intake-xarray and it's design and purpose. It seems to me that some of the complication with intake-xarray is the integration with fsspec and ultimately whether the library dealing with I/O can handle fsspec filelike objects. 

In the case of rioxarray / xr.open_rasterio. rasterio.open() does not know what to do with fsspec objects so we just pass urls around. and unless i'm missing something there is no need for fsspec?

Reply:
fsspec might be useful for expanding glob-strings; also, in some cases the best route is to have fsspec make local copies and pass the paths of those. It's the case where the xarray backend has it's own internal URL handling that's thorny. fsspec might still be useful for expanding globs, but neither file-likes or local copies are appropriate.

Reply:
thanks for the feedback @martindurant !

> It's the case where the xarray backend has it's own internal URL handling that's thorny. 

yes, and with the backend refactor ongoing it's especially confusing for me to understand. there is a proposal in rioxarray to handle fsspec objects https://github.com/corteva/rioxarray/issues/246 , so maybe eventually more xarray backends will support this...

> fsspec might still be useful for expanding globs, but neither file-likes or local copies are appropriate.

i think this is another case where intake-stac is a bit weird compared to other intake plugins. since STAC is already a catalog and we're basically translating to a new catalog type, i don't there there is a case where someone would manually write or save an intake yml version. And consequently there won't be any glob patterns in the URLs... 

in the absence of fsspec integrations the key goals of intake-stac in my mind are:

1. provide a common interface to load data (.to_dask()) regardless of file type (.jp2, .tif, .nc)... 
2. provide convenience functions for complicated loading of many assets (arguably this could also be implemented as an open_mfrasterio() in rioxarray, but having the STAC metadata allows us to take shortcuts or do things otherwise impossible with just filenames 
3. tap into the holoviews system for visualization